sdks/java/core/src/main/java/org/apache/beam/sdk/util/FlexibleBackoff.java

```diff
@@ -0,0 +1,139 @@
+  private static final Duration DEFAULT_MAX_BACKOFF = Duration.standardDays(1000);
+  private static final int DEFAULT_MAX_ATTEMPTS = Integer.MAX_VALUE;
+  private static final Duration DEFAULT_MAX_CUM_BACKOFF = Duration.standardDays(1000);
+  // Customization of this backoff.
+  private final double exponent;
+  private final Duration initialBackoff;
+  private final Duration maxBackoff;
+  private final Duration maxCumulativeBackoff;
+  private final int maxAttempts;
+  // Current state
+  private Duration currentCumulativeBackoff;
+  private int currentAttempt;
+
+  public static FlexibleBackoff of() {
+    return new FlexibleBackoff(
+        DEFAULT_EXPONENT,
+        DEFAULT_MIN_BACKOFF, DEFAULT_MAX_BACKOFF, DEFAULT_MAX_CUM_BACKOFF,
+        DEFAULT_MAX_ATTEMPTS);
+  }
+
+  public FlexibleBackoff withExponent(double exponent) {
+    checkArgument(exponent > 0, "exponent %s must be greater than 0", exponent);
+    return new FlexibleBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxAttempts);
+  }
+
+  public FlexibleBackoff withInitialBackoff(Duration initialBackoff) {
+    checkArgument(
+        initialBackoff.getMillis() > 0,
+        "initialBackoff %s must be at least 1 millisecond",
+        initialBackoff);
+    return new FlexibleBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxAttempts);
+  }
+
+  public FlexibleBackoff withMaxBackoff(Duration maxBackoff) {
+    checkArgument(
+        maxBackoff.getMillis() > 0,
+        "maxBackoff %s must be at least 1 millisecond",
+        maxBackoff);
+    return new FlexibleBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxAttempts);
+  }
+
+  public FlexibleBackoff withMaxCumulativeBackoff(Duration maxCumulativeBackoff) {
+    checkArgument(maxCumulativeBackoff.isLongerThan(Duration.ZERO),
+        "maxCumulativeBackoff %s must be at least 1 millisecond", maxCumulativeBackoff);
+    return new FlexibleBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxAttempts);
+  }
+
+  public FlexibleBackoff withMaxAttempts(int maxAttempts) {
+    checkArgument(maxAttempts >= 1, "maxAttempts %s must be at least 1", maxAttempts);
+    return new FlexibleBackoff(
+        exponent, initialBackoff, maxBackoff, maxCumulativeBackoff, maxAttempts);
+  }
+
+  @Override
+  public void reset() {
+    currentAttempt = 0;
+    currentCumulativeBackoff = Duration.ZERO;
+  }
+
+  @Override
+  public long nextBackOffMillis() {
+    // Maximum number of attempts reached.
+    if (currentAttempt >= maxAttempts) {
+      return BackOff.STOP;
+    }
+    // Maximum cumulative backoff reached.
+    if (currentCumulativeBackoff.equals(maxCumulativeBackoff)) {
+      return BackOff.STOP;
+    }
+
+    double currentIntervalMillis =
+        Math.min(
+            initialBackoff.getMillis() * Math.pow(exponent, currentAttempt),
```

Comment:
(If github showed branching, this is branched from the `IntervalBounded` code I'm deleting)